aws/resource_aws_budget.go

```diff
@@ -0,0 +1,303 @@
+	d.SetId(*budget.BudgetName)
+	return resourceAwsBudgetUpdate(d, meta)
+}
+
+func resourceAwsBudgetRead(d *schema.ResourceData, meta interface{}) error {
+	budgetName := d.Id()
+	describeBudgetOutput, err := describeBudget(budgetName, meta)
+	if isBudgetNotFoundException(err) {
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("describe budget failed: %v", err)
+	}
+
+	if _, ok := d.GetOk("name"); ok {
+		d.Set("name", describeBudgetOutput.Budget.BudgetName)
+	}
+
+	d.Set("budget_type", describeBudgetOutput.Budget.BudgetType)
+	d.Set("limit_amount", describeBudgetOutput.Budget.BudgetLimit.Amount)
+	d.Set("limit_unit", describeBudgetOutput.Budget.BudgetLimit.Unit)
+	d.Set("include_credit", describeBudgetOutput.Budget.CostTypes.IncludeCredit)
+	d.Set("include_other_subscription", describeBudgetOutput.Budget.CostTypes.IncludeOtherSubscription)
+	d.Set("include_recurring", describeBudgetOutput.Budget.CostTypes.IncludeRecurring)
+	d.Set("include_refund", describeBudgetOutput.Budget.CostTypes.IncludeRefund)
+	d.Set("include_subscription", describeBudgetOutput.Budget.CostTypes.IncludeSubscription)
+	d.Set("include_support", describeBudgetOutput.Budget.CostTypes.IncludeSupport)
+	d.Set("include_tax", describeBudgetOutput.Budget.CostTypes.IncludeTax)
+	d.Set("include_upfront", describeBudgetOutput.Budget.CostTypes.IncludeUpfront)
+	d.Set("use_blended", describeBudgetOutput.Budget.CostTypes.UseBlended)
+	d.Set("time_period_start", describeBudgetOutput.Budget.TimePeriod.Start)
+	d.Set("time_period_end", describeBudgetOutput.Budget.TimePeriod.End)
+	d.Set("time_unit", describeBudgetOutput.Budget.TimeUnit)
+	d.Set("cost_filters", describeBudgetOutput.Budget.CostFilters)
+	return nil
+}
+
+func resourceAwsBudgetUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AWSClient).budgetconn
+	accountID := meta.(*AWSClient).accountid
+	budget, err := newBudget(d)
+	if err != nil {
+		return fmt.Errorf("could not create budget: %v", err)
+	}
+
+	updateBudgetInput := new(budgets.UpdateBudgetInput)
+	updateBudgetInput.SetAccountId(accountID)
+	updateBudgetInput.SetNewBudget(budget)
+	_, err = client.UpdateBudget(updateBudgetInput)
+	if err != nil {
+		return fmt.Errorf("updaate budget failed: %v", err)
+	}
+
+	return resourceAwsBudgetRead(d, meta)
+}
+
+func resourceAwsBudgetDelete(d *schema.ResourceData, meta interface{}) error {
+	budgetName := d.Id()
+	if !budgetExists(budgetName, meta) {
+		log.Printf("[INFO] budget %s could not be found. skipping delete.", d.Id())
+		return nil
+	}
+
+	client := meta.(*AWSClient).budgetconn
+	accountID := meta.(*AWSClient).accountid
+	deleteBudgetInput := new(budgets.DeleteBudgetInput)
+	deleteBudgetInput.SetBudgetName(budgetName)
+	deleteBudgetInput.SetAccountId(accountID)
+	_, err := client.DeleteBudget(deleteBudgetInput)
+	if err != nil {
+		return fmt.Errorf("delete budget failed: %v", err)
+	}
+
+	return nil
+}
+
+func newBudget(d *schema.ResourceData) (*budgets.Budget, error) {
```

Comment:
Function names are global across the provider currently. We generally name functions like these `expandBudgetsBudget` to denote we're "expanding" the HCL configuration into a certain SDK data type.

Preferably, this would be split into its constituent parts as well (depending on resource schema nesting):
* `func expandBudgetsBudget()`
* `func expandBudgetsCostTypes(m map[string]interface{}) (*budgets.CostTypes, error)`